lib/web_ui/lib/src/engine/bitmap_canvas.dart

```diff
@@ -285,37 +296,107 @@ class BitmapCanvas extends EngineCanvas {
     _canvasPool.clipPath(path);
   }
 
+  /// Whether drawing operation should use DOM node instead of Canvas.
+  ///
+  /// - Perspective transforms are not supported by canvas and require
+  ///   DOM to render correctly.
+  /// - Pictures typically have large rect/rounded rectangles as background
+  ///   prefer DOM if canvas has not been allocated yet.
+  bool _useDomForRendering(SurfacePaintData paint) =>
+      _contains3dTransform || (_canvasPool._canvas == null
+          && paint.maskFilter == null && paint.shader == null &&
+          paint.style != ui.PaintingStyle.stroke);
+
   @override
   void drawColor(ui.Color color, ui.BlendMode blendMode) {
-    _canvasPool.drawColor(color, blendMode);
+    final SurfacePaintData paintData = SurfacePaintData()
+      ..color = color
+      ..blendMode = blendMode;
+    if (_useDomForRendering(paintData)) {
+      drawRect(ui.Rect.fromLTWH(0, 0, _bounds.width, _bounds.height),
+        paintData);
+    } else {
+      _canvasPool.drawColor(color, blendMode);
+    }
   }
 
   @override
   void drawLine(ui.Offset p1, ui.Offset p2, SurfacePaintData paint) {
-    _setUpPaint(paint);
-    _canvasPool.strokeLine(p1, p2);
-    _tearDownPaint();
+    if (_useDomForRendering(paint)) {
+      final SurfacePath path = SurfacePath()
+          ..moveTo(p1.dx, p1.dy)
+          ..lineTo(p2.dx, p2.dy);
+      drawPath(path, paint);
+    } else {
+      _setUpPaint(paint);
+      _canvasPool.strokeLine(p1, p2);
+      _tearDownPaint();
+    }
   }
 
   @override
   void drawPaint(SurfacePaintData paint) {
-    _setUpPaint(paint);
-    _canvasPool.fill();
-    _tearDownPaint();
+    if (_useDomForRendering(paint)) {
+      drawRect(ui.Rect.fromLTWH(0, 0, _bounds.width, _bounds.height), paint);
```

Comment:
Same comment here: will it cover the whole picture?

Reply:
Done.